File: 141/linked-list-cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
function hasCycle(head) {
    let hash = new Set();
    while (head != null) {
        if (hash.has(head)) {
            return true;
        } else {
            hash.add(head);
        }
        head = head.next;
    }
    return false;
}

File: 141/LinkedListCycle.java
/**
 * LinkedListCycle
 */
public class LinkedListCycle {

    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null)
            return false;
        
        ListNode slow = head, fast = head.next;
        while (fast != null && fast.next != null) {
            if (slow == fast) return true;
            else {
                slow = slow.next;
                fast = fast.next.next;
            }
        }
        return false;
    }

    class ListNode {
        int val;
        ListNode next;
        
        ListNode(int x) {
            val = x;
            next = null;
        }
    }
}